File: frameworks/python/fastapi/Implementation/payloads.py
"""The committed payloads, read from the directory RB_PAYLOADS names as each worker imports the
application, so a missing or broken file stops the boot rather than failing a request. The parsed
models are kept and serialised on every request."""
from dataclasses import dataclass
from pathlib import Path

# rb:wiring json.*
from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel


# rb:wiring json.*
class Camel(BaseModel):
    """A model whose JSON names are camelCase, as every name in the corpus is. FastAPI serializes
    a declared return type with Pydantic, straight to JSON bytes."""

    model_config = ConfigDict(alias_generator=to_camel, validate_by_name=True, serialize_by_alias=True)


class Item(Camel):
    """One row of items.large, and of every payload made from it."""

    id: int
    name: str
    category: str
    price_cents: int
    in_stock: bool


class Payload(Camel):
    """items.small, items.medium or items.large."""

    size: str
    count: int
    items: list[Item]
# rb:end


class Vary(Camel):
    """The values each vary row is keyed on, by header."""

    one: dict[str, list[str]]
    many: dict[str, list[str]]


class CacheSettings(Camel):
    capacity: int
    ttl_seconds: int
    vary: Vary


class CorsSettings(Camel):
    origin: str
    method: str
    header: str
    max_age_seconds: int


class Settings(Camel):
    """The values the framework configures itself from, as settings.json holds them."""

    token: str
    wrong_token: str
    stale_etag: str
    cache: CacheSettings
    cors: CorsSettings


@dataclass(frozen=True)
class Payloads:
    directory: Path
    small: Payload
    medium: Payload
    large: Payload
    settings: Settings
    rows: dict[int, Item]

    def row(self, id: int) -> Item | None:
        """The row of items.large with this id, or None when there is none."""
        return self.rows.get(id)


def load(directory: str) -> Payloads:
    root = Path(directory).resolve()
    large = Payload.model_validate_json((root / "items.large.json").read_bytes())
    return Payloads(
        directory=root,
        small=Payload.model_validate_json((root / "items.small.json").read_bytes()),
        medium=Payload.model_validate_json((root / "items.medium.json").read_bytes()),
        large=large,
        settings=Settings.model_validate_json((root / "settings.json").read_bytes()),
        rows={row.id: row for row in large.items},
    )

File: frameworks/java/javalin/Implementation/src/main/java/implementation/Payloads.java
package implementation;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * The committed payloads, read from the directory RB_PAYLOADS names before the server starts, so
 * a missing or broken file stops the boot rather than failing a request. The parsed objects are
 * kept and serialised on every request.
 */
public final class Payloads {

    private final Path directory;

    private final Payload small;

    private final Payload medium;

    private final Payload large;

    private final Settings settings;

    private final Map<Integer, Item> rows;

    private Payloads(Path directory) throws IOException {
        ObjectMapper json = new ObjectMapper();
        this.directory = directory.toAbsolutePath().normalize();
        this.small = read(json, "items.small.json", Payload.class);
        this.medium = read(json, "items.medium.json", Payload.class);
        this.large = read(json, "items.large.json", Payload.class);
        this.settings = read(json, "settings.json", Settings.class);
        this.rows = large.items().stream().collect(Collectors.toUnmodifiableMap(Item::id, Function.identity()));
    }

    public static Payloads load(Path directory) throws IOException {
        return new Payloads(directory);
    }

    public Path directory() {
        return directory;
    }

    public Payload small() {
        return small;
    }

    public Payload medium() {
        return medium;
    }

    public Payload large() {
        return large;
    }

    public Settings settings() {
        return settings;
    }

    /** The row of items.large with this id, or null when there is none. */
    public Item row(int id) {
        return rows.get(id);
    }

    private <T> T read(ObjectMapper json, String file, Class<T> type) throws IOException {
        return json.readValue(Files.readAllBytes(directory.resolve(file)), type);
    }
}
